File: backend/app/schemas/run_of_show.py
"""Run-of-Show Pydantic schemas."""

import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field


class RunOfShowItemCreate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=200)]
    description: str | None = None
    scheduled_time: datetime | None = None
    donor_visible: bool = False
    auctioneer_visible: bool = True
    display_order: int | None = None


class RunOfShowItemUpdate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=200)] | None = None
    description: str | None = None
    scheduled_time: datetime | None = None
    donor_visible: bool | None = None
    auctioneer_visible: bool | None = None
    display_order: int | None = None


class RunOfShowItemResponse(BaseModel):
    id: uuid.UUID
    event_id: uuid.UUID
    title: str
    description: str | None = None
    scheduled_time: datetime | None = None
    donor_visible: bool
    auctioneer_visible: bool
    is_complete: bool
    completed_at: datetime | None = None
    display_order: int
    has_notification: bool = False
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class RunOfShowResponse(BaseModel):
    items: list[RunOfShowItemResponse]
    total_count: int
    completed_count: int
    next_item: RunOfShowItemResponse | None = None
    event_start_time: datetime | None = None


class RunOfShowReorderRequest(BaseModel):
    item_ids: list[uuid.UUID]


class RunOfShowTemplateItemResponse(BaseModel):
    id: uuid.UUID
    title: str
    description: str | None = None
    offset_minutes: int
    donor_visible_default: bool
    auctioneer_visible_default: bool
    display_order: int
    model_config = ConfigDict(from_attributes=True)


class RunOfShowTemplateResponse(BaseModel):
    id: uuid.UUID
    npo_id: uuid.UUID | None = None
    name: str
    is_system_default: bool
    item_count: int = 0
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class RunOfShowTemplateDetailResponse(BaseModel):
    id: uuid.UUID
    npo_id: uuid.UUID | None = None
    name: str
    is_system_default: bool
    items: list[RunOfShowTemplateItemResponse]
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class RunOfShowTemplateItemCreate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=200)]
    description: str | None = None
    offset_minutes: Annotated[int, Field(ge=0)] = 0
    donor_visible_default: bool = True
    auctioneer_visible_default: bool = True
    display_order: int | None = None


class RunOfShowTemplateItemUpdate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=200)] | None = None
    description: str | None = None
    offset_minutes: Annotated[int, Field(ge=0)] | None = None
    donor_visible_default: bool | None = None
    auctioneer_visible_default: bool | None = None
    display_order: int | None = None


class SaveAsTemplateRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=200)]


class ApplyTemplateRequest(BaseModel):
    template_id: uuid.UUID
    confirm_replace: bool = False


class ApplyTemplateResponse(BaseModel):
    replaced: bool
    items_created: int


class RosNotificationCreate(BaseModel):
    message_body: Annotated[str, Field(min_length=1)]
    recipient_type: str  # "donors", "auctioneer", "all_attendees"


class RosNotificationResponse(BaseModel):
    id: uuid.UUID
    ros_item_id: uuid.UUID
    message_body: str
    recipient_type: str
    scheduled_at: datetime
    delivery_status: str
    celery_task_id: str | None = None
    delivered_at: datetime | None = None
    failure_reason: str | None = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

File: frontend/fundrbolt-admin/src/types/run-of-show.ts
export interface RunOfShowItem {
  id: string
  event_id: string
  title: string
  description: string | null
  scheduled_time: string // ISO datetime
  donor_visible: boolean
  auctioneer_visible: boolean
  is_complete: boolean
  completed_at: string | null
  display_order: number
  notifications: RosNotification[]
  created_by: string
  created_at: string
  updated_at: string
}

export interface RunOfShowResponse {
  items: RunOfShowItem[]
  total_count: number
  completed_count: number
  next_item: RunOfShowItem | null
  event_start_time: string | null
}

export interface RunOfShowItemCreate {
  title: string
  description?: string | null
  scheduled_time: string
  donor_visible?: boolean
  auctioneer_visible?: boolean
  display_order?: number | null
}

export interface RunOfShowItemUpdate {
  title?: string
  description?: string | null
  scheduled_time?: string
  donor_visible?: boolean
  auctioneer_visible?: boolean
  display_order?: number | null
}

export interface RunOfShowReorderRequest {
  item_ids: string[]
}

export interface RunOfShowTemplate {
  id: string
  npo_id: string | null
  name: string
  is_system_default: boolean
  item_count: number
  created_by: string | null
  created_at: string
  updated_at: string
}

export interface RunOfShowTemplateItem {
  id: string
  title: string
  description: string | null
  offset_minutes: number
  donor_visible_default: boolean
  auctioneer_visible_default: boolean
  display_order: number
}

export interface RunOfShowTemplateDetail {
  id: string
  npo_id: string | null
  name: string
  is_system_default: boolean
  items: RunOfShowTemplateItem[]
  created_by: string | null
  created_at: string
  updated_at: string
}

export interface SaveAsTemplateRequest {
  name: string
}

export interface ApplyTemplateRequest {
  template_id: string
  confirm_replace?: boolean
}

export interface ApplyTemplateResponse {
  replaced: boolean
  items_created: number
}

export interface RosNotification {
  id: string
  ros_item_id: string
  message_body: string
  recipient_type: 'donors' | 'auctioneer' | 'all_attendees'
  minutes_before: number
  scheduled_at: string
  delivery_status: 'pending' | 'delivered' | 'failed' | 'cancelled'
  celery_task_id: string | null
  delivered_at: string | null
  failure_reason: string | null
  created_at: string
  updated_at: string
}

export interface RosNotificationCreate {
  message_body: string
  recipient_type: 'donors' | 'auctioneer' | 'all_attendees'
  minutes_before?: number
}
